### ISSUE TITLE
Multiple versions of react

### ISSUE THREAD

**Issue by nirchernia** [namespace-ee/react-calendar-timeline#35]
Hey, thanks for the awesome module!

I am having an issue with multiple versions of react, although it may likely be an issue on my end. I thought you guys might be able to help.

First of all my package json looks like this:

```
{
  "name": "monitor",
  "peerDependencies": {
    "react": "^15.1.0"
  },
  "version": "0.0.0",
  "private": true,
  "scripts": {
    "start": "node ./bin/www"
  },
  "dependencies": {
    "body-parser": "~1.13.2",
    "flux": "~2.1.1",
    "gulp": "~3.9.1",
    "gulp-babel": "^6.1.2",
    "gulp-browserify": "~0.5.1",
    "gulp-concat": "~2.6.0",
    "gulp-sass": "^2.3.2",
    "lodash": "^4.13.1",
    "moment": "~2.13.0",
    "interact.js": "^1.2.6",
    "react-addons-css-transition-group": "~15.0.2",
    "react-big-calendar": "~0.9.11",
    "react-calendar-timeline": "^0.8.1",
    "react-datetime": "~2.1.0",
    "react-dom": "^15.1.0",
    // more that are unrelated
  },
  "devDependencies": {
    // unrelated
  },
  "main": "app.js",
  "author": "",
  "license": "ISC",
  "description": ""
}
```

As you can see I have react 15.1.0 as a peerDependency. When I had a problem with another module giving off multiple versions of react, I was told to supply react as a `peerDependency` instead of a regular dependency and that fixed my issue

What is strange is while requiring using react-calendar-timeline, I am receiving the famous multiple versions  of react problem again: 

> ScheduleMain.js:43829 Uncaught Invariant Violation: addComponentAsRefTo(...): Only a ReactOwner can have refs. You might be adding a ref to a component that was not created inside a component's `render` method, or you have multiple copies of React loaded (details: https://fb.me/react-refs-must-have-owner).

At the top of my ScheduleMain.js I import `react` and `react-dom`. I do this for all my react scripts
Not entirely sure why this is the case since it seems like this lib uses react 15.1.0:

```
[root@monitor]$ npm ls react-calendar-timeline
/var/www//monitor
├── UNMET PEER DEPENDENCY react@^15.1.0
├─┬ react-calendar-timeline@0.8.1
│ ├── UNMET PEER DEPENDENCY phantomjs@>=1.9
│ └── UNMET PEER DEPENDENCY react@15.1.0
└── UNMET PEER DEPENDENCY react-dom@15.1.0
```

I am using NPM 3.
Does anyone know why this is happening?

**Comment by mariusandra** [comment]
Hey,
I can't pinpoint your issue, but it feels wrong to list `react` as a `peerDependency` in your app. What package made you do so? My guess is that this package has `react` as a regular `dependency` and that break things. Probably if you fix that package and put `react` back as a `dependency`, the problems will go away.

**Comment by nirchernia** [comment]
@mariusandra 

The package was searchkit: https://github.com/namespace-ee/react-calendar-timeline/blob/master/package.json

And it seems they do the same thing that this package is doing, aren't they?

react-calendar-timeline:
```

  "dependencies": {
    "interact.js": "^1.2.6",
    "moment": "^2.11.1",
    "react": "^15.0.1",
    "react-dom": "^15.0.1",
    "react-pure-render": "^1.0.2"
  },
  "peerDependencies": {
    "react": "^0.14.8 || ^15.0.1",
    "react-dom": "^0.14.8 || ^15.0.1",
    "moment": "*",
    "interact.js": "*"
  },

```


**searchkit:**
```
  "dependencies": {
    "axios": "^0.9.1",
    "bem-cn": "^1.2.2",
    "es6-promise": "^3.1.2",
    "history": "^2.1.1",
    "lodash": "^4.6.1",
    "qs": "ssetem/qs",
    "rc-slider": "^3.3.2",
    "react": "^0.14.7",
    "react-addons-update": "^0.14.7"
  },
```

Is it because they didn't provide it as a peer dependency also? I also don't really understand why react-calendar-timeline has both peer and regular dep for react

**Comment by mariusandra** [comment]
Quoting a [stack overflow answer](http://stackoverflow.com/a/34645112/624476) on this topic:

> TL;DR: peerDependencies are for dependencies that are exposed to (and expected to be used by) the consuming code, as opposed to 'private' dependencies that are not exposed and are only an implementation detail.

If a plugin (e.g. searchkit or react-calednar-timeline) defines something as a peer dependency, it expects that your app already includes the thing as a dependency, and the module will just use whatever was included. If the plugin does not specify that it is a peer dependency, it will install it itself. 

In your case this means that searchkit will install its own version of React. That's why the suggestion for you to define it as a peerdependency in your project made sense - searchkit was the "one source of truth" when it came to React and the version it defined was used.

This is not how it should be. Every included plugin must declare react as a peer dependency. Otherwise they're doing it wrong. Searchkit looks awesome, but the way they handle dependencies is breaking your app.

My suggestion unless you have a better alternative: fork searchkit, add a peerDependency of React to their package.json and use that... or find a way to write this elasticsearch logic without searchkit. The latter is probably a more sustainable approach as your app grows.

Cheers and good luck! :)

**Comment by nirchernia** [comment]
Thanks for the thorough response @mariusandra. It seems as though there is a bit of a philosophical disagreement between handling peerDependencies and Dependencies as a whole. Regardless, this helped me greatly!
